src/kbmod/search/logging.h

```diff
@@ -0,0 +1,221 @@
+  // so just pass the message and level onward and let it be resolved in Python
+  class PyLogger : public Logger {
+  private:
+    py::handle pylogger;
+
+  public:
+    PyLogger(py::handle logger) :
+      Logger(logger.attr("name").cast<std::string>()),
+      pylogger(logger)
+    {}
+
+    virtual void log(std::string level, std::string msg){
+      for (char& ch : level) ch = std::tolower(ch);
+      pylogger.attr(level.c_str())(msg);
+    }
+  };
+#endif // Py_PYTHON_H
+
+
+  // Glorified std::cout. Unlike the PyLogger, CoutLogger is configurable.
+  // Configuration should contain `level`, `datefmt` and `format` keys:
+  // - `level`:  LogLevel enum value, the minimal level that is printed
+  // - `datefmt`: timestamp template usable with `strftime`
+  // - `format`: log format template, too much commitment atm, not supported
+  class CoutLogger : public Logger{
+  private:
+    // datefmt %Y-%m expands to YYYY-MM etc., 2x its size sounds safe enough but
+    // with abuse I guess this could overflow?
+    std::string fmt_time(){
+      time_t now;
+      time(&now);
+      char buf[2*sizeof(config["datefmt"])] = {0};
+      strftime(buf, sizeof buf, config["datefmt"].c_str(), gmtime(&now));
+      return std::string(buf);
+    }
+
+  public:
+    CoutLogger(std::string name, sdict config) :
+      Logger(name, config)
+    {}
+
+    virtual void log(std::string level, std::string msg){
+      if (level_threshold <= StringToLogLevel[level])
+        std::cout << "[" << fmt_time() << " " << level << " " << name << "] " << msg << std::endl;
+    }
+  };
+
+
+  // The singleton keeping the registry of all registered Loggers and their
+  // default configuration. Use `logger` to access the singleton instance,
+  // getLoger to get an existing logger or to create a new default logger.
+  // No good examples, but let's say a non-stdout C++ only logger is required,
+  // use getLogger<LoggerCls>(name, conf) to register and get the instance of it
+  class Logging{
+  private:
+    static sdict default_config;
+    std::map<std::string, Logger*> registry;
+    static Logging* instance;
+
+    Logging(){}
+    ~Logging(){}
+
+  public:
+    Logging(Logging &other) = delete;
+    void operator=(const Logging &) = delete;
+
+    // get the singleton instance
+    static Logging* logger(){
+      if(instance == nullptr)
+        instance = new Logging();
+      return instance;
+    }
+
+    static void setConfig(sdict config){
+      default_config = config;
+    }
+
+    template<class LoggerCls>
+    static Logger* getLogger(std::string name, sdict config={}){
+      if (instance == nullptr)
+        instance = new Logging();
+
+      // if key not found use default setup
+      if (instance->registry.find(name) == instance->registry.end()) {
+        sdict tmpconf = config.size() != 0 ? config : instance->default_config;
+        instance->registry[name] = new LoggerCls(name, tmpconf);
```

Comment:
I was very sloppy with the memory management, but that's mostly because I didn't understand it. Not so much the basics of malloc and dealloc, but like when is the default destructor called, is `py:handle` reference managed, why does [this official pybind11 documentation snippet](https://pybind11.readthedocs.io/en/stable/advanced/classes.html#non-public-destructors) not work and so on. 
I went back through all of it and I think this has it covered now, but I don't actually know:

- Fix singleton object lifetime by wrapping it into a static method
- use `py::object` instead of `py::handle` to wrap a Python Logger instance
  - this respects the refcount tracking, unlike before. Why we didn't
    get an error at the end of every program execution - I don't know.
- add destructors, call them for Logger in the registry, let the
  default per-field destructor handle the rest.
- Add virtual destructor to Logger class itself,
  - Deleting a derived class object using a pointer of base class type
    that has a non-virtual destructor results in undefined behavior.
- Fix the potential memory leak on time-stamp via put_time instead
  of mallocing until strftime does not throw an error.

Technically, we should be calling `make_unique` instead of instantiating a unique pointer directly - since apparently if our object fails during construction, that's undefined behavior also - but I can't make pybind work any other way without getting double allocated, double deallocated errors.

Point out if you can see anything else.